File: src/main/java/komu/blunt/types/ClassInstance.java
package komu.blunt.types;

import komu.blunt.types.checker.Substitution;

import java.util.List;
import java.util.Set;

import static com.google.common.base.Preconditions.checkNotNull;

public final class ClassInstance implements Types<ClassInstance> {
    final Qualified<Predicate> qual;

    public ClassInstance(Qualified<Predicate> qual) {
        this.qual = checkNotNull(qual);
    }
    
    public List<Predicate> getPredicates() {
        return qual.predicates;
    }

    public Predicate getPredicate() {
        return qual.value;
    }

    @Override
    public void addTypeVariables(Set<TypeVariable> variables) {
        qual.addTypeVariables(variables);
    }

    @Override
    public ClassInstance apply(Substitution substitution) {
        return new ClassInstance(qual.apply(substitution));
    }

    @Override
    public String toString() {
        return qual.toString();
    }
}

File: src/main/kotlin/komu/blunt/types/ClassInstance.kt
package komu.blunt.types

import komu.blunt.types.checker.Substitution

class ClassInstance(val qual: Qualified<Predicate>) : Types<ClassInstance> {

    val predicates: List<Predicate>
        get() = qual.predicates

    val predicate: Predicate
        get() = qual.value

    override fun typeVars(): Sequence<Type.Var> = qual.typeVars()

    override fun apply(substitution: Substitution): ClassInstance =
        ClassInstance(qual.apply(substitution))

    override fun toString() =
        qual.toString()
}
